README.md
# dxfuse: a FUSE filesystem for dnanexus

A filesystem that allows users access to the DNAnexus storage system.

[![Build Status](https://travis-ci.org/dnanexus/dxfuse.svg?branch=master)](https://travis-ci.org/dnanexus/dxfuse)

**NOTE: This is a project in its beta stage. We are using it on cloud workers, however, it may be run from any machine with a network connection and a DNAx account.**

The code uses the [FUSE](https://bazil.org/fuse/)
library, implemented in [golang](https://golang.org). The DNAnexus
storage system is not POSIX compilant. It holds not just files and
directories, but also records, databases, applets, and workflows. It
allows things that are not POSIX, for example:
1. Files in a directory can have the same name
2. A filename can include slashes
3. A file and a directory may share a name

To fit these names into a POSIX compliant filesystem, as FUSE and
Linux require, files are moved to avoid name collisions. For example,
if file `foo.txt` has three versions in directory `bar`, dxfuse will
present the following Unix directory structure:

```
bar/
    foo.txt
    1/foo.txt
    2/foo.txt
```

The directories `1` and `2` are new, and do not exist in the
project. If a file contains a slash, it is replaced with a triple
underscore. As a rule, directories are not moved, nor are their
characters modified. However, if a directory name contains a slash, it
is dropped, and a warning is emitted to the log.

dxfuse approximates a normal POSIX filesystem, but does not always have the same semantics. For example:
1. Metadata like last access time are not supported
2. Directories have approximate create/modify times. This is because DNAx does not keep such attributes for directories.

There are several limitations currently:
- Primarily intended for Linux, but can be used on OSX
- Limits directories to 10,000 elements
- Updates to the project emanating from other machines are not reflected locally
- Rename does not allow removing the target file or directory. This is because this cannot be
  done automatically by dnanexus.
- Does not support hard links

Updates to files are batched and asynchronously applied to the cloud
object system. For example, if `foo.txt` is updated, the changes will
not be immediately visible to another user looking at the platform
object directly. Because platform files are immutable, even a minor
modification requires rewriting the entire file, creating a new
version. This is an inherent limitation, making file update
inefficient.

## Implementation

The implementation uses an [sqlite](https://www.sqlite.org/index.html)
database, located on `/var/dxuse/metadata.db`. It stores files and
directories in tables, indexed to speed up common queries.

Load on the DNAx API servers and the cloud object system is carefully controlled. Bulk calls
are used to describe data objects, and the number of parallel IO requests is bounded.

dxfuse operations can sometimes be slow, for example, if the server is
slow to respond, or has been temporarily shut down (503 mode). This
may cause the filesystem to lose its interactive feel. Running it on a
cloud worker reduces network latency significantly, and is the way it
is used in the product. Running on a local, non cloud machine, runs
the risk of network choppiness.

Bandwidth when streaming a file is close to the dx-toolkit, but may be a
little bit lower. The following table shows performance across several
instance types. The benchmark was *how many seconds does it take to
download a file of size X?* The lower the number, the better. The two
download methods were (1) `dx cat`, and (2) `cat` from a dxfuse mount point.

| instance type   | dx cat (seconds) | dxfuse cat (seconds) | file size |
| ----            | ----             | ---                  |  ----     |
| mem1\_ssd1\_v2\_x4|	         207 |	219                 |       17G |
| mem1\_ssd1\_v2\_x4|	          66 |      	         77 |      5.9G |
| mem1\_ssd1\_v2\_x4|		6|	4 | 705M|
| mem1\_ssd1\_v2\_x4|		3|	3 | 285M|
| | | | |
| mem1\_ssd1\_v2\_x16|	57|	49 | 17G|
| mem1\_ssd1\_v2\_x16|		22|	24 | 5.9G|
| mem1\_ssd1\_v2\_x16|		3|	3 | 705M|
| mem1\_ssd1\_v2\_x16|		2|	1 | 285M|
| | | | |
| mem3\_ssd1\_v2\_x32|		52|	51 | 17G|
| mem3\_ssd1\_v2\_x32|		20	| 15 | 5.9G |
| mem3\_ssd1\_v2\_x32|		4|	2 | 705M|
| mem3\_ssd1\_v2\_x32|		2|	1 | 285M|

Creating new files, and uploading them to the platform is
significantly slower when using dxfuse. This is mostly a FUSE problem,
where IOs are handed over from user-space, to the kernel, and then to
the dxfuse server in 128KB chunks. This is fine for small files, but
becomes slow for large files. Here is a comparison between (1) `dx
upload FILE`, and (2) `cp FILE to fuse, sync fuse`.

| instance type      | dx upload (seconds) | dxfuse upload (seconds) | file size |
| ----               | ----                | ---                  |  ----     |
| mem1\_ssd1\_v2\_x4 |	11 |	99 | 705M |
| mem1_ssd1\_v2\_x4  |	5  |	24 | 285M |
| mem1\_ssd1\_v2\_x16 |	10 |	67 | 705M |
| mem1\_ssd1\_v2\_x16 |	5  |	17 | 285M |
| mem3\_ssd1\_v2\_x32 |	9  |	34 | 705M |
| mem3\_ssd1\_v2\_x32 |	6  |	12 | 285M |

# Building

To build the code from source, you'll need, at the very least, the `go` and `git` tools.
install dependencies:
```
go get github.com/google/subcommands
go get golang.org/x/sync/semaphore
go install github.com/google/subcommands
go get github.com/dnanexus/dxda
go install github.com/dnanexus/dxda
go install github.com/dnanexus/dxda/cmd/dx-download-agent
```

Assuming the go directory is `/go`, clone the code with:
```
cd /go/src/github.com/dnanexus/dxfuse
git clone git@github.com:dnanexus/dxfuse.git
```

Build the code:
```
go build -o /go/bin/dxfuse /go/src/github.com/dnanexus/dxfuse/cli/main.go
```

# Usage

To mount a dnanexus project `mammals` on local directory `/home/jonas/foo` do:
```
sudo -E dxfuse -uid $(id -u) -gid $(id -g) /home/jonas/foo mammals
```

The bootstrap process has some asynchrony, so it could take it a
second two to start up. It spawns a separate process for the filesystem
server, waits for it to start, and exits. To get more information, use
the `verbose` flag. Debugging output is written to the log, which is
placed at `/var/log/dxfuse.log`. The maximal verbosity level is 2.

```
sudo -E dxfuse -verbose 1 MOUNT-POINT PROJECT-NAME
```

Project ids can be used instead of project names. To mount several projects, say, `mammals`, `fish`, and `birds`, do:
```
sudo -E dxfuse /home/jonas/foo mammals fish birds
```

This will create the directory hierarchy:
```
/home/jonas/foo
              |_ mammals
              |_ fish
              |_ birds
```

Note that files may be hard linked from several projects. These will appear as a single inode with
a link count greater than one.

To stop the dxfuse process do:
```
sudo umount MOUNT-POINT
```

There are situations where you want the background process to
synchronously update all modified and newly created files. For example, before shutting down a machine,
or unmounting the filesystem. This can be done by issuing the command:
```
$ sudo dxfuse -sync
```

## Extended attributes (xattrs)

DNXa data objects have properties and tags, these are exposed as POSIX extended attributes. Xattrs can be read, written, and removed. The package we use here is `attr`, it can installed with `sudo apt-get install attr` on Linux. On OSX the `xattr` package comes packaged with the base operating system, and can be used to the same effect.

DNAx tags and properties are prefixed. For example, if `zebra.txt` is a file then `attr -l zebra.txt` will print out all the tags, properties, and attributes that have no POSIX equivalent. These are split into three correspnding prefixes _tag_, _prop_, and _base_ all under the `user` Linux namespace.

Here `zebra.txt` has no properties or tags.
```
$ attr -l zebra.txt

base.state: closed
base.archivalState: live
base.id: file-xxxx
```

Add a property named `family` with value `mammal`
```
$ attr -s prop.family -V mammal zebra.txt
```

Add a tag `africa`
```
$ attr -s tag.africa -V XXX zebra.txt
```

Remove the `family` property:
```
$ attr -r prop.family zebra.txt
```

You cannot modify _base.*_ attributes, these are read-only. Currently, setting and deleting xattrs can be done only for files that are closed on the platform.

## Mac OS (OSX)

For OSX you will need to install [OSXFUSE](http://osxfuse.github.com/). Note that Your Milage May Vary (YMMV) on this platform, we are mostly focused on Linux.

# Common problems

If a project appears empty, or is missing files, it could be that the dnanexus token does not have permissions for it. Try to see if you can do `dx ls YOUR_PROJECT:`.

If you do not set the `uid` and `gid` options then creating hard links will fail on Linux. This is because it will fail the kernel's permissions check.

There is no natural match for DNAnexus applets and workflows, so they are presented as block devices. They do not behave like block devices, but the shell colors them differently from files and directories.

Mmap doesn't work all that well with FUSE ([stack overflow issue](https://stackoverflow.com/questions/46839807/mmap-no-such-device)). For example, trying to memory-map (mmap) a file with python causes an error.

```
>>> import mmap
>>> fd = open('/home/orodeh/MNT/dxfuse_test_data/README.md', 'r')
>>> mmap.mmap(fp.fileno(), 0, mmap.PROT_READ)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  OSError: [Errno 19] No such device
```

A workaround is to make the mapping private:

```
>>> import mmap
>>> fd = open('/home/orodeh/MNT/dxfuse_test_data/README.md', 'r')
>>> mmap.mmap(fd.fileno(), 0, prot=mmap.PROT_READ, flags=mmap.MAP_PRIVATE, offset=0)
>>> fd.readline()
```


RELEASE_NOTES.md
# Release Notes

## v0.15

- Improved command line invocation. You can now run:
```
$ /go/bin/dxfuse MOUNTPOINT PROJ1 PROJ2
```

instead of:

```
$ sudo -E /go/bin/dxfuse -uid $(id -u) -gid $(id -g) MOUNTPOINT PROJ1 PROJ2
```


## v0.14
Initializing the filesystem with a subprocess. The dxfuse program spawns a subprocess that runs
the actual filesystem. The main program waits for the subprocess to start and then returns the status.
If an initialization error occurs, it is reported synchronously. This obviates the need for shell scripts like:

```
$ /go/bin/dxfuse ... &
sleep 2
```

You can just write:
```
$ /go/bin/dxfuse
```

Rename works. There is one important limitation, which is, that you cannot rename onto an existing file.
This is because dnanexus does not guaranty that remove + rename is atomic. For example, if you have
directory `A` with files `foo` and `bar`, then this works:
```
$ mv A/foo A/X
```

But this does not:
```
$ mv A/foo A/bar
```

A less important limitation is that faux directories cannot be moved. Such directories arise when a platform
directory holds files with the same name. For example:

```
$ dx ls proj-xxxx:/A

Chicago.txt : file-FgXbQ2j04J92g9b87V2zPF59
Chicago.txt : file-FgXbPG004J90ZB7q62p191vZ
NewYork.txt : file-FgXbPb004J980yX164B4KPFy
NewYork.txt : file-FgXbPJQ04J94B2xg637jQJqz
```

In the filesystem, these are represented like this:
```
A
├── 1
│   ├── Chicago.txt
│   └── NewYork.txt
├── Chicago.txt
└── NewYork.txt
```

The directory `1` is not real, it does not have a backend folder on the platform. Therefore, it cannot be moved or
modified. It may be erased, if all of its files are removed.

Hard links are supported. There are backend limitations to their use. You can make a link to a file
in project `A` from a file in project `B`. However, you will get an error if you link to a file in the same project. This is not allowed with the dnanexus `clone` operation.

Streaming speed on a cloud worker is quite good, generally on par with `dx cat`. On a remote machine streaming still works, but uses much smaller buffer sizes. Large HTTP requests tend to incur long timeouts causing streaming to slow down to a crawl.

## v0.13 21-Nov-2019
- Additional operations supported: create directory, remove directory, file unlink

## v0.12  12-Nov-2019
- Migrated to the [jacobsa](https://github.com/jacobsa/fuse) golang fuse filesystem package
- File creation supported. Such files are uploaded to the platform once they are closed. They also
become read-only on close. This mimics the DNAx behavior, which is that files are immutable.

## v0.11  26-Sep-2019
- Initial release


util.go
package dxfuse

import (
	"database/sql"
	"fmt"
	"log"
	"os"
	"sync"
	"time"

	"github.com/dnanexus/dxda"
	"github.com/hashicorp/go-retryablehttp"

	"github.com/jacobsa/fuse/fuseutil"
	"github.com/jacobsa/fuse/fuseops"
)

const (
	ConfigFile =         "/tmp/dxfuse_config.json"
	CreatedFilesDir =    "/var/dxfuse/created_files"
	DatabaseFile       = "/var/dxfuse/metadata.db"
	LogFile            = "/var/log/dxfuse.log"
)

const (
	HttpClientPoolSize = 4
	MaxDirSize         = 10 * 1000
	MaxNumFileHandles  = 1000 * 1000
	NumRetriesDefault  = 3
	Version            = "v0.15"
)
const (
	InodeInvalid       = 0
	InodeRoot          = fuseops.RootInodeID  // This is an OS constant
)
const (
	KiB                   = 1024
	MiB                   = 1024 * KiB
	GiB                   = 1024 * MiB
)
const (
	// It turns out that in order for regular users to be able to create file,
	// we need to use 777 permissions for directories.
	dirReadOnlyMode = 0555 | os.ModeDir
	dirReadWriteMode = 0777 | os.ModeDir
	fileReadOnlyMode = 0444
)

// A URL generated with the /file-xxxx/download API call, that is
// used to download file ranges.
type DxDownloadURL struct {
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers"`
}

type Options struct {
	ReadOnly            bool   `json:"readOnly"`
	Verbose             bool   `json:"verbose"`
	VerboseLevel        int    `json:"verboseLevel"`
	Uid                 uint32 `json:"uid"`
	Gid                 uint32 `json:"gid"`
}


type Filesys struct {
	// inherit empty implementations for all the filesystem
	// methods we do not implement
	fuseutil.NotImplementedFileSystem

	// configuration information for accessing dnanexus servers
	dxEnv dxda.DXEnvironment

	// various options
	options Options

	// A file holding a sqlite3 database with all the files and
	// directories collected thus far.
	dbFullPath string

	// Lock for protecting shared access to the database
	mutex sync.Mutex

	// a pool of http clients, for short requests, such as file creation,
	// or file describe.
	httpClientPool    chan(*retryablehttp.Client)

	// metadata database
	mdb *MetadataDb

	// prefetch state for all files
	pgs *PrefetchGlobalState

	// background upload state
	fugs *FileUploadGlobalState

	// all open files
	fhTable map[fuseops.HandleID]*FileHandle
	fhFreeList []fuseops.HandleID

	// all open directories
	dhTable map[fuseops.HandleID]*DirHandle
	dhFreeList []fuseops.HandleID

	nonce *Nonce
	tmpFileCounter uint64

	// is the the system shutting down (unmounting)
	shutdownCalled bool
}

// A node is a generalization over files and directories
type Node interface {
	GetInode() fuseops.InodeID
	GetAttrs() fuseops.InodeAttributes
}


// directories
type Dir struct {
	Parent      string  // the parent directory, used for debugging
	Dname       string  // This is the last part of the full path
	FullPath    string // combine parent and dname, then normalize
	Inode       int64
	Ctime       time.Time // DNAx does not record times per directory.
	Mtime       time.Time // we use the project creation time, and mtime as an approximation.
	Mode        os.FileMode  // uint32
	Uid         uint32
	Gid         uint32

	// extra information, used internally
	ProjId      string
	ProjFolder  string
	Populated   bool

	// is this a faux dir?
	faux        bool
}

func (d Dir) GetAttrs() (a fuseops.InodeAttributes) {
	a.Size = 4096
	a.Nlink = 1
	a.Mtime = d.Mtime
	a.Ctime = d.Ctime
	a.Mode = os.ModeDir | d.Mode
	a.Crtime = d.Ctime
	a.Uid = d.Uid
	a.Gid = d.Gid
	return
}

func (d Dir) GetInode() fuseops.InodeID {
	return fuseops.InodeID(d.Inode)
}


// Kinds of files
const (
	FK_Regular = 10
	FK_Symlink = 11
	FK_Applet = 12
	FK_Workflow = 13
	FK_Record = 14
	FK_Database = 15
	FK_Other = 16
)

// A Unix file can stand for any DNAx data object. For example, it could be a workflow or an applet.
// We distinguish between them based on the Id (file-xxxx, applet-xxxx, workflow-xxxx, ...).
type File struct {
	Kind       int     // Kind of object this is
	Id         string  // Required to build a download URL
	ProjId     string  // Note: this could be a container
	Name       string
	Size       int64
	Inode      int64
	Ctime      time.Time
	Mtime      time.Time
	Mode       os.FileMode  // uint32
	Nlink      int
	Uid        uint32
	Gid        uint32

	// for a symlink, it holds the path.
	// For a regular file, a path to a local copy (if any).
	InlineData string
}

func (f File) GetAttrs() (a fuseops.InodeAttributes) {
	a.Size = uint64(f.Size)
	a.Nlink = uint32(f.Nlink)
	a.Mtime = f.Mtime
	a.Ctime = f.Ctime
	a.Mode = f.Mode
	a.Crtime = f.Ctime
	a.Uid = f.Uid
	a.Gid = f.Gid
	return
}

func (f File) GetInode() fuseops.InodeID {
	return fuseops.InodeID(f.Inode)
}


// Files can be opened in read-only mode, or read-write mode.
const (
	RO_Remote = 1     // read only file that is on the cloud
	RW_File = 2       // newly created file
	RO_LocalCopy = 3  // read only file that has a local copy
)

type FileHandle struct {
	fKind int
	f File

	// URL used for downloading file ranges.
	// Used for read-only files.
	url *DxDownloadURL

	// Local file copy, may be empty.
	localPath *string

	// 1. Used for reading from an immutable local copy
	// 2. Used for writing to newly created files.
	fd *os.File
}

type DirHandle struct {
	d Dir
	entries []fuseutil.Dirent
}

// A handle used when operating on a filesystem
// operation. We normally need a transaction and an http client.
type OpHandle struct {
	httpClient *retryablehttp.Client
	txn        *sql.Tx
	err         error
}

func (oph *OpHandle) RecordError(err error) error {
	oph.err = err
	return err
}

// Utility functions

func MaxInt64(x, y int64) int64 {
    if x < y {
        return y
    }
    return x
}

func MinInt64(x, y int64) int64 {
    if x > y {
        return y
    }
    return x
}

func MaxInt(x, y int) int {
    if x < y {
        return y
    }
    return x
}

func MinInt(x, y int) int {
    if x > y {
        return y
    }
    return x
}

// convert time in seconds since 1-Jan 1970, to the equivalent
// golang structure
func SecondsToTime(t int64) time.Time {
	return time.Unix(t, 0)
}

func Time2string(t time.Time) string {
	return fmt.Sprintf("%02d:%02d:%02d.%03d", t.Hour(), t.Minute(), t.Second(), t.Nanosecond()/1000000)
}

// add a timestamp and module name, to a log message
func LogMsg(moduleName string, a string, args ...interface{}) {
	msg := fmt.Sprintf(a, args...)
	now := time.Now()
	log.Printf("%s %s: %s", Time2string(now), moduleName, msg)
}


//
// 1024   => 1KB
// 10240  => 10KB
// 1100000 => 1MB
func BytesToString(numBytes int64) string {
	byteModifier := []string {"B", "KB", "MB", "GB", "TB", "EB", "ZB" }

	digits := make([]int, 0)
	for numBytes > 0 {
		d := int(numBytes % 1024)
		digits = append(digits, d)
		numBytes = numBytes / 1024
	}

	// which digit is the most significant?
	msd := len(digits)
	if msd >= len(byteModifier) {
		// This number is so large that we don't have a modifier
		// for it.
		return fmt.Sprintf("%dB", numBytes)
	}
	return fmt.Sprintf("%d%s", digits[msd], byteModifier[msd])
}


cli/main.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"os/user"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/jacobsa/fuse"
	"github.com/jacobsa/fuse/fuseutil"

	// The dxda package has the get-environment code
	"github.com/dnanexus/dxda"
	"github.com/dnanexus/dxfuse"
)

type Config struct {
	Mountpoint string             `json:"mountpoint"`
	DxEnv      dxda.DXEnvironment `json:"dxEnv"`
	Options    dxfuse.Options     `json:"options"`
	Manifest   dxfuse.Manifest    `json:"manifest"`
}

var progName = filepath.Base(os.Args[0])

func usage() {
	fmt.Fprintf(os.Stderr, "usage:\n")
	fmt.Fprintf(os.Stderr, "    %s [options] MOUNTPOINT PROJECT1 PROJECT2 ...\n", progName)
	fmt.Fprintf(os.Stderr, "    %s [options] MOUNTPOINT manifest.json\n", progName)
	fmt.Fprintf(os.Stderr, "options:\n")
	flag.PrintDefaults()
	fmt.Fprintf(os.Stderr, "\n")
	fmt.Fprintf(os.Stderr, "A project can be specified by its ID or name. The manifest is a JSON\n")
	fmt.Fprintf(os.Stderr, "file describing the initial filesystem structure.\n")
}

var (
	debugFuseFlag = flag.Bool("debugFuse", false, "Tap into FUSE debugging information")
	help = flag.Bool("help", false, "display program options")
	readOnly = flag.Bool("readOnly", false, "mount the filesystem in read-only mode")
	verbose = flag.Int("verbose", 0, "Enable verbose debugging")
	version = flag.Bool("version", false, "Print the version and exit")
)

func lookupProject(dxEnv *dxda.DXEnvironment, projectIdOrName string) (string, error) {
	if strings.HasPrefix(projectIdOrName, "project-") {
		// This is a project ID
		return projectIdOrName, nil
	}

	// This is a project name, describe it, and
	// return the project-id.
	return dxfuse.DxFindProject(context.TODO(), dxEnv, projectIdOrName)
}

func initLog() *os.File {
	// Redirect the log output to a file
	f, err := os.OpenFile(dxfuse.LogFile, os.O_RDWR | os.O_CREATE | os.O_APPEND | os.O_TRUNC, 0666)
	if err != nil {
		log.Fatalf("error opening file: %v", err)
	}
	log.SetOutput(f)
	log.SetFlags(0)
	log.SetPrefix(progName + ": ")

	return f
}

func getUidGid() (uint32,uint32) {
	// This is current the root user, because the program is run under
	// sudo privileges. The "user" variable is used only if we don't
	// get command line uid/gid.
	user, err := user.Current()
	if err != nil {
		panic(err)
	}

	// get the user ID
	uid, err := strconv.Atoi(user.Uid)
	if err != nil {
		panic(err)
	}

	// get the group ID
	gid, err := strconv.Atoi(user.Gid)
	if err != nil {
		panic(err)
	}

	return uint32(uid),uint32(gid)
}

// Mount the filesystem:
//  - setup the debug log to the FUSE kernel log (I think)
//  - mount as read-only
func fsDaemon(
	mountpoint string,
	dxEnv dxda.DXEnvironment,
	manifest dxfuse.Manifest,
	options dxfuse.Options) error {

	fsys, err := dxfuse.NewDxfuse(dxEnv, manifest, options)
	if err != nil {
		return err
	}
	server := fuseutil.NewFileSystemServer(fsys)

	// initialize the log file
	logf := initLog()
	defer logf.Close()
	logger := log.New(logf, "dxfuse: ", log.Flags())

	logger.Printf("starting fsDaemon")
	mountOptions := make(map[string]string)

	// Allow users other than root access the filesystem
	mountOptions["allow_other"] = ""

	// capture debug output from the FUSE subsystem
	var fuse_logger *log.Logger
	if *debugFuseFlag {
		fuse_logger = log.New(logf, "fuse_debug: ", log.Flags())
	}

	logger.Printf("building config")

	// Fuse mount
	cfg := &fuse.MountConfig{
		FSName : "dxfuse",
		ErrorLogger : logger,
		DebugLogger : fuse_logger,
		DisableWritebackCaching : true,
		Options : mountOptions,
	}

	logger.Printf("mounting dxfuse")
	os.Stderr.WriteString("Ready")
	os.Stderr.Close()
	mfs, err := fuse.Mount(mountpoint, server, cfg)
	if err != nil {
		logger.Printf(err.Error())
	}

	// Wait for it to be unmounted. This happens only after
	// all requests have been served.
	if err = mfs.Join(context.Background()); err != nil {
		logger.Fatalf("Join: %v", err)
	}

	// shutdown the filesystem
	fsys.Shutdown()

	return nil
}

func waitForReady(readyReader *os.File, c chan string) {
	status := make([]byte, 1000)
	_, err := readyReader.Read(status)
	if err != nil {
		log.Printf("Reading from ready pipe: %v", err)
		os.Exit(1)
	}
	c <- string(status)
}

func parseCmdLineArgs() Config {
	if *version {
		// print the version and exit
		fmt.Println(dxfuse.Version)
		os.Exit(0)
	}
	if *help {
		usage()
		os.Exit(0)
	}

	numArgs := flag.NArg()
	if numArgs < 2 {
		usage()
		os.Exit(2)
	}
	mountpoint := flag.Arg(0)

	uid,gid := getUidGid()

	options := dxfuse.Options{
		ReadOnly: *readOnly,
		Verbose : *verbose > 0,
		VerboseLevel : *verbose,
		Uid : uid,
		Gid : gid,
	}

	dxEnv, _, err := dxda.GetDxEnvironment()
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}

	if dxEnv.DxJobId == "" {
		fmt.Fprintf(os.Stderr, `
Warning: running outside a worker. Dxfuse is currently engineered to
operate inside a cloud worker. The system depends on a good network
connection to the DNAnexus servers, and to the backing store, which is
S3 or Azure. Without such connectivity, some operations may take a
long time, causing operating system timeouts to expire. This can
result in the filesystem freezing, or being unmounted.
`)
	}

	return Config{
		Mountpoint : mountpoint,
		DxEnv : dxEnv,
		Options : options,
	}
}

func parseManifest(cfg Config) (*dxfuse.Manifest, error) {
	numArgs := flag.NArg()

	// distinguish between the case of a manifest, and a list of projects.
	if numArgs == 2 && strings.HasSuffix(flag.Arg(1), ".json") {
		p := flag.Arg(1)
		log.Printf("Provided with a manifest, reading from %s", p)
		manifest, err := dxfuse.ReadManifest(p)
		if err != nil {
			return nil, err
		}
		if err := manifest.FillInMissingFields(context.TODO(), cfg.DxEnv); err != nil {
			return nil, err
		}
		return manifest, nil
	} else {
		// process the project inputs, and convert to an array of verified
		// project IDs
		var projectIds []string
		for i := 1; i < numArgs; i++ {
			projectIdOrName := flag.Arg(i)
			projId, err := lookupProject(&cfg.DxEnv, projectIdOrName)
			if err != nil {
				return nil, err
			}
			if projId == "" {
				return nil, fmt.Errorf("no project with name %s", projectIdOrName)
			}
			projectIds = append(projectIds, projId)
		}

		manifest, err := dxfuse.MakeManifestFromProjectIds(context.TODO(), cfg.DxEnv, projectIds)
		if err != nil {
			return nil, err
		}
		return manifest, nil
	}
}

func configFileRead(filename string) (*Config, error) {
	payload, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	var cfg Config
	if err := json.Unmarshal(payload, &cfg); err != nil {
		return nil, err
	}
	return &cfg, nil
}

func configFileWrite(cfg Config, filename string) error {
	payload, err := json.Marshal(cfg)
	if err != nil {
		return err
	}
	if err := ioutil.WriteFile(filename, payload, 0444); err != nil {
		return err
	}
	return nil
}

// This runs with root permissions
func runAsDaemonWithRootPermisions() {
	// read the configuration file
	cfg, err := configFileRead(dxfuse.ConfigFile)
	if err != nil {
		os.Stderr.WriteString("Error, configuration file not specified")
		os.Stderr.Close()
		os.Exit(1)
	}

	err = fsDaemon(cfg.Mountpoint, cfg.DxEnv, cfg.Manifest, cfg.Options)
	if err != nil {
		os.Stderr.WriteString(err.Error())
		os.Stderr.Close()
		os.Exit(1)
	}
	return
}

func main() {
	// parse command line options
	flag.Usage = usage
	flag.Parse()

	if flag.NArg() == 1 {
		// running without any command line arguments
		runAsDaemonWithRootPermisions()
		return
	}

	// normal user.
	// create the configuration
	cfg := parseCmdLineArgs()

	manifest, err := parseManifest(cfg)
	if err != nil {
		fmt.Fprintf(os.Stderr, err.Error())
		fmt.Fprintf(os.Stderr, "Could not parse the manifest")
		os.Exit(1)
	}
	if manifest == nil {
		fmt.Fprintf(os.Stderr, "Manifest is empty")
		os.Exit(1)
	}
	cfg.Manifest = *manifest

	// write to a file
	if err := configFileWrite(cfg, dxfuse.ConfigFile); err != nil {
		fmt.Fprintf(os.Stderr, err.Error())
		fmt.Fprintf(os.Stderr, "Could not write configuration file into %s\n", dxfuse.ConfigFile)
		os.Exit(1)
	}

	// Set up a pipe for the "ready" status.
	errorReader, errorWriter, err := os.Pipe()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Pipe: %v", err)
		os.Exit(1)
	}
	defer errorWriter.Close()
	defer errorReader.Close()

	// Mount in a subprocess, and wait for the filesystem to start.
	// If there is an error, report it. Otherwise, return after the filesystem
	// is mounted and useable.
	//
	progPath, err := exec.LookPath(os.Args[0])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: couldn't find program %s", os.Args[0])
		os.Exit(1)
	}
	mountCmd := exec.Command(progPath)
	mountCmd.Stderr = errorWriter

	// Start the command.
	fmt.Println("starting fs daemon")
	if err := mountCmd.Start(); err != nil {
		fmt.Fprintf(os.Stderr, "failed to start filesystem daemon: %v\n", err)
		os.Exit(1)
	}

	// Wait for the tool to say the file system is ready. In parallel, watch for
	// the tool to fail.
	fmt.Println("wait for ready")
	readyChan := make(chan string, 1)
	go waitForReady(errorReader, readyChan)

	status := <-readyChan
	status = strings.ToLower(status)
	if strings.HasPrefix(status, "ready") {
		fmt.Println("Daemon started successfully")
		return
	}

	fmt.Println("There was an error starting the daemon")
	fmt.Println(status)
	os.Exit(1)
}
